source/simulator.hpp

```diff
@@ -10,42 +15,174 @@
+        return m_switches[name];
+    }
 
-    std::shared_ptr<Flow> add_flow(std::shared_ptr<ISender> sender,
-                                   std::shared_ptr<IReceiver> recever,
-                                   Size a_packet_size,
-                                   Time a_delay_between_packets,
-                                   std::uint32_t packets_to_send);
+    std::shared_ptr<TFlow> add_flow(std::shared_ptr<ISender> sender,
+                                    std::shared_ptr<IReceiver> receiver,
+                                    Size packet_size,
+                                    Time delay_between_packets,
+                                    std::uint32_t packets_to_send) {
+        auto flow =
+            std::make_shared<Flow>(sender, receiver, packet_size,
+                                   delay_between_packets, packets_to_send);
+        m_flows.emplace_back(flow);
+        return flow;
+    }
 
     void add_link(std::shared_ptr<IRoutingDevice> a_from,
                   std::shared_ptr<IRoutingDevice> a_to,
-                  std::uint32_t a_speed_mbps, Time a_delay);
+                  std::uint32_t a_speed_mbps, Time a_delay) {
+        auto link = std::make_shared<Link>(a_from, a_to, a_speed_mbps, a_delay);
+        m_links.emplace_back(link);
+        a_from->add_outlink(link);
+        a_to->add_inlink(link);
+    }
 
-    std::vector<std::shared_ptr<IRoutingDevice>> get_devices() const;
+    std::vector<std::shared_ptr<IRoutingDevice>> get_devices() const {
+        std::vector<std::shared_ptr<IRoutingDevice>> result;
+        std::transform(m_senders.begin(), m_senders.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+
+        std::transform(m_receivers.begin(), m_receivers.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+
+        std::transform(m_switches.begin(), m_switches.end(),
+                       std::back_inserter(result),
+                       [](const auto& pair) { return pair.second; });
+        return result;
+    }
 
     // Clear all events in the Scheduler
-    void clear();
+    void clear() {}
     // BFS to update the routing table
-    void recalculate_paths();
+    void recalculate_paths() {
+        for (auto src_device : get_devices()) {
+            routing_table_t routing_table = bfs(src_device);
+            for (auto [dest_device, link] : routing_table) {
+                src_device->update_routing_table(dest_device, link);
+            }
+        }
+    }
     // Create a Stop event at a_stop_time and start simulation
-    void start(Time a_stop_time);
+    void start(Time a_stop_time) {
+        recalculate_paths();
+        Scheduler::get_instance().add(std::make_unique<Stop>(a_stop_time));
+        constexpr Time start_time = 0;
+
+        for (std::shared_ptr<IFlow> flow : m_flows) {
```

Comment:
`m_flows` contains `Flow` objects. I think it ok yo use `auto` here:
```suggestion
        for (flow : m_flows) {
```